pydantic_ai_graph/pydantic_ai_graph/graph.py

```diff
@@ -2,169 +2,232 @@
-        import httpx
+        else:
+            for _, next_node_id in sorted((node_order[node_id], node_id) for node_id in node_def.next_node_ids):
+                lines.append(f'  {node_id} --> {next_node_id}')
+        if node_def.returns_end:
+            lines.append(f'  {node_id} --> END')
 
-        code_base64 = base64.b64encode(self.mermaid_code().encode()).decode()
+    return '\n'.join(lines)
 
-        response = httpx.get(f'https://mermaid.ink/img/{code_base64}', params=mermaid_ink_params)
-        response.raise_for_status()
-        return response.content
 
-    def mermaid_save(self, path: Path | str, mermaid_ink_params: dict[str, str | int] | None = None) -> None:
-        image_data = self.mermaid_image(mermaid_ink_params)
-        Path(path).write_bytes(image_data)
+def mermaid_image(
+    graph: Graph[Any, Any],
+    start: StartNodeProtocol[..., Any, Any] | tuple[StartNodeProtocol[..., Any, Any], ...] = (),
+    mermaid_ink_params: dict[str, str | int] | None = None,
+) -> bytes:
+    import httpx
 
-    def _check(self):
-        bad_edges: dict[str, list[str]] = {}
-        for node in self.nodes.values():
-            node_bad_edges = node.next_node_ids - self.nodes.keys()
-            for bad_edge in node_bad_edges:
-                bad_edges.setdefault(bad_edge, []).append(f'"{node.node_id}"')
+    code_base64 = base64.b64encode(mermaid_code(graph, start).encode()).decode()
 
-        if bad_edges:
-            bad_edges_list = [f'"{k}" is referenced by {_utils.comma_and(v)}' for k, v in bad_edges.items()]
-            if len(bad_edges_list) == 1:
-                raise ValueError(f'{bad_edges_list[0]} but not included in the graph.')
-            else:
-                b = '\n'.join(f' {be}' for be in bad_edges_list)
-                raise ValueError(f'Nodes are referenced in the graph but not included in the graph:\n{b}')
+    response = httpx.get(f'https://mermaid.ink/img/{code_base64}', params=mermaid_ink_params)
+    response.raise_for_status()
+    return response.content
 
 
-def get_parent_namespace(frame: types.FrameType | None) -> dict[str, Any] | None:
-    """Attempt to get the namespace where the graph was defined.
-
-    If the graph is defined with generics `Graph[a, b]` then another frame is inserted, and we have to skip that
-    to get the correct namespace.
-    """
-    if frame is not None:
-        if back := frame.f_back:
-            if back.f_code.co_filename.endswith('/typing.py'):
-                return get_parent_namespace(back)
-            else:
-                return back.f_locals
+def mermaid_save(
+    path: Path | str,
+    graph: Graph[Any, Any],
+    start: StartNodeProtocol[..., Any, Any] | tuple[StartNodeProtocol[..., Any, Any], ...] = (),
+    mermaid_ink_params: dict[str, str | int] | None = None,
+) -> None:
+    # TODO: do something with the path file extension, e.g. error if it's incompatible, or use it to specify a param
```

Comment:
I think we remove `mermaid_ink_params` and just list out the params, then we can respect `type` from https://mermaid.ink/